wip/services/exec/store.go

```diff
@@ -18,55 +17,128 @@ func NewPostgresStore() Store {
 	return &postgresStore{}
 }
 
-func (p postgresStore) Create(project string, exec types.Exec) error {
-	var command []string
-	var commitID, gitRemoteURL sql.NullString
-
-	if exec.Command != nil {
-		command = exec.Command
-	}
-	if exec.CommitID != nil {
-		commitID = sql.NullString{String: *exec.CommitID, Valid: true}
-	}
-	if exec.GitURL != nil {
-		gitRemoteURL = sql.NullString{String: *exec.GitURL, Valid: true}
+func (p postgresStore) Create(ctx context.Context, project string, exec types.Exec) error {
+	if project == "" {
+		return fmt.Errorf("an Exec must be attached to a project")
 	}
-	bid := sql.NullString{}
-	if exec.BuildID != nil {
-		bid = sql.NullString{String: *exec.BuildID, Valid: true}
+
+	// every exec should be created with a public key
+	publicKeys := types.FilterKeysWithPublicKey(exec.Keys)
+	if len(publicKeys) > 1 || len(publicKeys) == 0 {
+		return fmt.Errorf("an Exec must be created with one and only one SSH public key")
 	}
 	if exec.Name == "" {
-		exec.Name = random.GenerateRandomPhrase(4, "-")
+		return fmt.Errorf("an Exec must be named")
 	}
+
 	spec, err := json.Marshal(&exec.Spec)
 	if err != nil {
 		return fmt.Errorf("failed to marshal spec to JSON: %w", err)
 	}
+
 	metadata, err := json.Marshal(&types.NodeMetadataV1{})
 	if err != nil {
 		return fmt.Errorf("failed to marshal metadata to JSON: %w", err)
 	}
 
-	dbp := db.ExecCreateParams{
-		ID:           exec.ID,
-		CreatedBy:    exec.CreatedBy,
-		ProjectID:    project,
-		Region:       exec.Region,
-		Name:         exec.Name,
-		Spec:         spec,
-		Metadata:     metadata,
-		CommitID:     commitID,
-		GitRemoteUrl: gitRemoteURL,
-		Command:      command,
-		BuildID:      bid,
-		Image:        exec.Image,
-		Provider:     exec.Provider.String(),
-	}
-
-	if err := db.Q.ExecCreate(context.Background(), dbp); err != nil {
-		// TODO: parse db errors: project not found, ssh key not found,
+	err = createSSHKeys(ctx, exec.CreatedBy, publicKeys)
+	if err != nil {
+		return err
+	}
+	err = createExec(project, spec, metadata, exec)
+	if err != nil {
+		return err
+	}
+	keys, err := getSSHKeysByPublicKey(ctx, exec.CreatedBy, types.GetPublicKeys(publicKeys))
+	if err != nil {
+		return err
+	}
+	err = addSSHKeyToExec(ctx, exec, keys)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createExec(projectID string, spec []byte, metadata []byte, exec types.Exec) error {
```

Comment:
I don't think this needs a separate function. This Just adds additional places you can make an error by passing in `[]byte` arrarys

Reply:
Yeah, good catch